[MVC - Step 1] 미스터코 미션 제출합니다.

eunsukko opened the pull request:
써머님 안녕하세요..! 

어떻게 진행하는게 좋을까 막막하기도 했고
집중이 안되는 상황들이 있어서 많이 늦어졌습니다 ㅠ!

일단 요구사항 작성해서 제출합니다.
특이한 부분, 이런식으로 해봤으면 하는 부분 있으시면 편하게 말씀해주셔요..!
(제가 생각하기에 부족한 부분들을 다듬어보고 있겠습니다..!)
감사합니다..!!

Conatuseus committed:
[feat] java.lang.Class를 사용해 클래스 정보 확인

Conatuseus committed:
[feat] test로 시작하는 메서드 실행 확인

Conatuseus committed:
[feat] MyTest annotation이 있는 메서드 실행  확인

Conatuseus committed:
[feat] private field에 값 할당 가능여부 확인

Conatuseus committed:
[feat] 인자를 가진 생성자의 인스턴스 생성 확인

Conatuseus committed:
[feat] component scan

Conatuseus committed:
[refactor] RequestMapping method 수정

Conatuseus committed:
[feat] find annotation

Conatuseus committed:
[feat] find annotation method

eunsukko committed:
[refactor] AnnotationHandlerMapping 에서 역할 추출

- ComponentScanner
	- 특정 어노테이션을 단 클래스를 찾아내는 역할
- ValueExtractor
	- 어노테이션으로 부터 해당하는 이름의 값을 추출하는 역할

eunsukko committed:
[refactor] AnnotationHandlerMapping 리팩토링(불필요한 부분 제거)

eunsukko committed:
[feat] ㅑ

eunsukko committed:
[refactor] HomeController 어노테이션 적용

- 점진적으로 Controller 를 바꿀 수 있도록
 	- DispatcherServlet 에서 mvc.asis, mvc.tobe 를 모두 지원

eunsukko committed:
[refactor] ProfileController 어노테이션 적용

eunsukko committed:
[refactor] UpdateFormUserController 어노테이션 적용

eunsukko committed:
[refactor] ForwadController 어노테이션 적용

eunsukko committed:
[refactor] CreateUserController 어노테이션 적용

eunsukko committed:
[refactor] 유저 관련 기능 UserController 에 모으기

eunsukko committed:
[refactor] 로그인, 로그아웃 기능 어노테이션 적용

eunsukko committed:
[refactor] 유저 수정 기능 UserController 내부로 이동

eunsukko committed:
[refactor] HandlerAdaptor 적용

- Handler 를 적용할 부분이 확장 가능하도록 적용

jihan805 commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/AnnotationHandlerMapping.java:
```diff
@@ -1,25 +1,82 @@
+        }
+    }
+
+    private void registerHandler(String basePackagePrefix) {
```
해당 클래스 깔끔하게 잘 구현하였어요! 
registerHandler가 하는 역할을 한번 더 분리할 수 있을 것 같은데, 메서드를 분리해보면 어떨까요?
혹은 ComponentScanner를 ControllerScanner로 변경해도 좋을 것 같아요!

eunsukko replied:
`@Controller` 가 달린 특정 클래스에 대해서,
내부 method 들을 등록하도록 쪼개보았습니다..! 
(말씀해주신 의도가 맞을까요?? ㅎ)

jihan805 commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/AnnotationHandlerMapping.java:
```diff
@@ -1,25 +1,82 @@
+            return handlerExecutions.get(handlerKey);
+        }
+
+        HandlerKey emptyMethodHandlerKey = new HandlerKey(request.getRequestURI(), null);
```
null을 사용하는 것이 좋은 방법일까요?

eunsukko replied:
이 부분은 RequestMethod.ALL 이 생기면서 함께 처리되었습니다..!

jihan805 commented on nextstep-mvc/src/main/java/nextstep/utils/ComponentScanner.java:
```diff
@@ -0,0 +1,43 @@
+import java.util.Map;
+import java.util.Set;
+
+public class ComponentScanner {
```
ComponentScanner 구현 👍

jihan805 commented on nextstep-mvc/src/test/java/nextstep/web/annotation/RequestMethodTest.java:
```diff
@@ -0,0 +1,7 @@
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class RequestMethodTest {
```
작성하지 않을 테스트는 삭제하는 것은 어떨까요?
테스트를 작성하는 것이 쉽지 않을 수 있지만 필요하다고 생각되는 부분들에 대해서도 고민해보고, 전체적으로 테스트 코드 작성에 대해 한번 더 검토해보면 어떨까요?

jihan805 commented on nextstep-mvc/src/main/java/nextstep/utils/ComponentScanner.java:
```diff
@@ -0,0 +1,43 @@
+        return instances;
+    }
+
+    public Set<Class<?>> scanAnnotatedClasses(Class<? extends Annotation> annotation) {
```
private 접근자를 사용하지 않은 특별한 이유가 있을까요?

eunsukko replied:
해당 메소드를 ComponentScanner 외부에서 사용했었습니다 (리팩토링 진행하다 보니 사용하는 곳이 ComponentScanner 내부만 남았네요)

다시생각해보니 해당 메소드를 외부에 제공하는 것도 좀 이상한 것 같습니다 (ComponentScanner 에서 제공해줄 메소드는 아닌 것 같네요)

jihan805 commented on nextstep-mvc/src/main/java/nextstep/mvc/DispatcherServlet.java:
```diff
@@ -30,28 +35,26 @@ public void init() throws ServletException {
-
-        RequestDispatcher rd = req.getRequestDispatcher(viewName);
-        rd.forward(req, resp);
+    protected void service(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
```
service 메서드가 하는일이 너무 많은것 같아요!
한단계 더 작게 쪼개 볼까요?

jihan805 commented on study/src/test/java/reflection/ReflectionTest.java:
```diff
@@ -5,6 +5,11 @@
+import java.lang.reflect.Modifier;
+import java.util.Arrays;
+import java.util.Date;
+import java.util.List;
```
logger를 사용하는 것도 좋지만 assert를 이용해 값을 비교하는 형식의 테스트를 작성해보면 어떨까요?

jihan805 commented on slipp/src/main/java/slipp/controller/HomeController.java:
```diff
@@ -1,15 +1,20 @@
-        req.setAttribute("users", DataBase.findAll());
-        return "home.jsp";
+@nextstep.web.annotation.Controller
+public class HomeController {
```
모든 controller를 어노테이션 기반으로 변경한것 같아요, 두 방식이 공존할 때 잘 동작하는지를 확인해보면 어떨까요?

eunsukko replied:
점진적으로 한다는게.. 조금 흥분되서 싹 다바꿔버렸습니다.. ㅠㅎ

HomeController 를 기존 방식으로 복구해놓았습니다!

테스트도 전체적으로 손봐야하는데,
공존할 때 검증할 수 있는 부분을 함께 살펴볼께요..!

jihan805 commented on slipp/src/main/java/slipp/AnnotationHandlerMapping.java:
```diff
@@ -0,0 +1,91 @@
+import java.util.Map;
+import java.util.Optional;
+
+public class AnnotationHandlerMapping implements HandlerMapping {
```
현재 AnnotationHandlerMapping 클래스가 slipp/mvc 패키지 아래 두개가 있는 것으로 보여져요. 확인 부탁드려요~!
더불어 어노테이션 기반의 처리를 담당하는 것은 프레임워크이 역할이에요. 어떤 패키지에 위치하는 것이 적절할까요?

eunsukko replied:
세심한 리뷰 감사합니다..!

말씀하신대로 프레임워크의 역할이 맞다고 생각이 들어서 mvc 패키지로 이동하였습니다.

추가로, `ManualHandlerMapping` 클래스 또한 프레임워크에서 지원해주면 좋을 것 같은 부분이여서 옮겼습니다. 
(url 과 컨트롤러 간의 매핑정보만 프레임워크 사용자가 제공해주면 된다고 생각했습니다! 사용자 편의를 위한 빌더를 제공하였습니다 :) )

jihan805 commented on slipp/src/main/java/slipp/AnnotationHandlerMapping.java:
```diff
@@ -0,0 +1,91 @@
+        RequestMapping requestMapping = method.getAnnotation(RequestMapping.class);
+
+        RequestMethod[] methods = requestMapping.method();
+        return new HandlerKey(requestMapping.value(), (methods == null) ? null : methods[0]);
```
요구사항 중 method가 설정되지 않은 경우 모든 HttpMethod를 지원하라는 기능이 있어요.
관련 요구사항에 대해서 확인 후, 테스트도 작성해보면 어떨까요?

eunsukko replied:
해당 요구사항을 놓치고 있었네요 ㅠ!

AnnotationHandlerMapping 은 등록할 handlerKey, handleExecution 을 생성해줍니다.
- `HandlerExecutionHandlerMapping' 이라는 클래스에게 '등록된 handler 를 찾는 역할' 을 위임합니다

이렇게 분리를 해보았더니 테스트 할 수 있는 부분이 늘어나네요.!
(클래스 역할이 살짝 애매한 것도 같은데.. 좀 더 고민이 필요한 것 같기도 합니다.. :))

jihan805 reviewed (changes requested):
안녕하세요 ~! 구현 잘하셨어요 💯 
몇가지 피드백 남겼으니 확인 후 머지할게요 :)

eunsukko committed:
[refactor] ManualHandlerMapping 구현을 nextstep-mvc 로 이동

- 왜 변경이 있었는가?
	- 애매한 AnnotationHandlerMapping 의 위치
		- mvc 프레임워크에 해당 내용이 있는게 맞다고 판단
		- ManualHandlerMapping 또한 같은 이유로 프레임워크에 있어야 한다고 판단

- 수정사항
	- 사용자 편의를 위한 빌더 추가

eunsukko committed:
[refactor] ComponentScanner.scanAnnotatedClasses 메소드 private 으로 변경

- 왜 변경(혹은 추가)이 있었나요?
	- ComponentScanner.scanAnnotatedClasses 가 ComponentScanner 에서 제공할 만한 기능은 아니라고 판단
		- 외부에 노출되지 않아야함

- 무엇이 변경되었나요?
	- ComponentScanner.scanAnnotatedClasses 를 private 으로 변경

eunsukko committed:
[feat] AnnotationHandlerMapping - RequestMethod.ALL 지원

- 왜 변경(혹은 추가)이 있었나요?
	- 'RequestMapping.method 에 아무 설정이 없을 경우 모든 메소드 타입을 지원' 요구사항 추가
- 무엇이 변경되었나요?
	- AnnotationHandlerMapping 에서 RequestMethod.ALL 적용
		- 해당 url 에 ALL 이 존재할 경우
			- RequestMethod 에 존재하는 모든 메소드 타입을 지원가능
		- 동일한 url 에 GET, ALL 이 존재할 경우
			- 구체적인 메소드인 GET 을 반환

eunsukko committed:
[refactor] 불필요한 테스트파일 삭제

eunsukko committed:
[refactor] DispatcherServlet 내부 역할 리팩토링

- 왜 변경(혹은 추가)이 있었나요?
	- DispatcherServlet.service() 가 많은 역할을 하고 있음

- 무엇이 변경되었나요?
	- service() 함수에선 동일한 정도로 추상화 된 메소드들이 보이도록
		- 흐름을 다루는 역할
	- adapter 에서 공통적으로 적용될 부분을 render 란 디폴트메소드로 추출

eunsukko committed:
[refactor] ReflectionTest 리팩토링

eunsukko commented:
피드백 반영해보았습니다..! 

전체적인 테스트도 좀 더 손보고 싶은데.. 조금 시간이 필요할 것 같아서 먼저 요청보냅니다..! (되는데로 함께 push 할께요..!)
감사합니다 :)

jihan805 commented on nextstep-mvc/src/main/java/nextstep/web/annotation/RequestMethod.java:
```diff
@@ -1,5 +1,5 @@
 
 public enum RequestMethod {
-    GET, HEAD, POST, PUT, PATCH, DELETE, OPTIONS, TRACE
+    GET, HEAD, POST, PUT, PATCH, DELETE, OPTIONS, TRACE, ALL
```
ALL 을 추가하는 것도 좋지만, ALL이 RequstMethod는 아니라고 생각해요 !
ALL 대신 method가 없는 경우 전체 method를 가지고와 등록해주는 구조로 변경하는 것이 조금 더 적합하지 않을까요?

eunsukko replied:
말씀해주신대로 ALL 이 RequestMethod 라고 할 수 없겠어요.

이렇게 했던 이유는 ALL 을 지원해야 하는 경우에 8개의 RequestMethod 를 등록해주어야하는게 낭비이지 않을까 였습니다. (mapping 에서 굳이.. 8개나 등록을 해줘야 할까??...란 생각에)

다시 생각해보면 ALL 을 사용할 경우 (즉 `@RequestMapping.method == 빈 값` 일 경우) 자체가 많지 않을 것도 같구요.

질문드리고 싶은 부분은
1. ALL을 사용하는 컨트롤러를 사용할 일이 많을까요?? (아직 많이 구현을 해보지 않아서 어떤 경우가 이런 예가 될지 감이 안오네요 ㅠ!)
    - 없다면 ALL 인 경우 전체 메소드를 등록해주어도 부담이 별로 없겠네요..!
2. ALL을 사용하는 경우가 많다면 다른 대안이 있을까요? 즉 모든 메소드를 등록하는게 부담이 있다면요?? (지금의 경우가 아니여도 왠지 mapping 을 해주는 곳들이라면 있을 법한 경우여서 궁금합니다..ㅎ)
    - 일단 제가 생각나는 방법은 밑에 적어보았습니다. (이 방법은 ALL 을 사용하는 곳이 많아질 경우, 비슷한 역할의 코드가 여러곳에 존재하게 될 것 같아서요..)

```
// before
public class HandlerKey {
    private String url;
    private RequestMethod requestMethod;
}

// after
public class HandlerKey {
    private String url;
    private Optional<RequestMethod> requestMethod;
}
```
이렇게 생각해보면  RequestMethod 를 사용하는 HandlerKey 에서
'RequestMethod 가 없는 경우(기존의 RequestMethod.ALL 을 사용하던)'를 책임지게 되는 것 같구요
(즉,  __`@RequestMapping` 으로 전달된 빈 method 정보를 '모든 타입의 RequestMethod'로 생각해줘__ 이 책임을 HandlerKey 가 가지는게 맞는것 같다는 생각이 듭니다.)

jihan805 commented on study/src/test/java/reflection/ReflectionTest.java:
```diff
@@ -3,41 +3,249 @@
+    @Test
+    public void clazz_getName() {
+        Class<Question> clazz = Question.class;
+        assertThat(clazz.getName()).isEqualTo("reflection.Question");
```
반영 💯 
테스트를 작성할때 "정~말" 부득이한 경우가 아니라면 logger를 사용하기보다는 assert를 사용하도록 해요 !!

jihan805 commented on nextstep-mvc/src/main/java/nextstep/mvc/handlermapping/AnnotationHandlerMapping.java:
```diff
@@ -0,0 +1,95 @@
+        }
+    }
+
+    private void registerHandlerFromMethods(Class<?> controllerClass, Object controller, HandlerExecutionHandlerMapping.Builder mappingBuilder) {
```
한 line의 length가 너무 긴것 같아요!!
https://google.github.io/styleguide/javaguide.html#s4.4-column-limit

jihan805 commented on nextstep-mvc/src/test/java/nextstep/mvc/handlermapping/HandlerExecutionHandlerMappingTest.java:
```diff
@@ -0,0 +1,116 @@
+    }
+
+    @Test
+    @DisplayName("RequestMethod.ALL 은 모든 종류의 RequestMethod 들을 통해 조회 가능")
```
display name을 사용하여 테스트 하고자 하는 의도를 명확하게 나타냄 👍 
이번 단계 여기서 머지할게요 :) 
추가/수정하고 싶은 테스트는 다음단계 이어서 진행하면서 푸쉬하면 같이 확인해 볼게요 !

jihan805 commented on slipp/src/main/java/slipp/ManualHandlerMapping.java:
```diff
@@ -1,42 +1,57 @@
 public class ManualHandlerMapping implements HandlerMapping {
-    private static final Logger logger = LoggerFactory.getLogger(DispatcherServlet.class);
-    private Map<String, Controller> mappings = new HashMap<>();
+    private static final Logger logger = LoggerFactory.getLogger(ManualHandlerMapping.class);
```
코멘트에 남겨주셔서 확인해보았는데 아직 옮겨지지 않은 것 같아요 :)

eunsukko replied:
앗... 세심한 리뷰 감사합니다 ㅠㅎ
옮겨놓았습니다!

jihan805 reviewed (approved):
안녕하세요~! 구현 & 피드백 반영 잘하셨어요 💯 
이번단계 머지 할게요, 간단한 코멘트 남겨두었으니 다음단계 진행하면서 같이 확인해주세요 !!
고생 많으셨어요  :)
